File: daemon/src/rpc/wire.ts
/**
 * NDJSON wire envelope per docs/tui-architecture.md §10.
 * Stub WorldSnapshot here; full shape filled in once Day 5+ wires the daemon-side
 * AgentEventSink and asset/layout modules. Clients are expected to be forward-compatible
 * with extra fields landing in later phases (`schemaVersion: 1` gates breaking changes).
 */

export const PROTO_VERSION = 1;

export type WireMessage = Req | Res | Evt | Hello | HelloAck;

export interface Req {
  kind: 'req';
  reqId: number;
  method: string;
  params: unknown;
}

export type Res =
  | { kind: 'res'; reqId: number; ok: true; data: unknown }
  | { kind: 'res'; reqId: number; ok: false; error: WireError };

export interface WireError {
  code: string;
  message: string;
}

export interface Evt {
  kind: 'evt';
  topic: string;
  seq: number;
  ts: number;
  data: unknown;
}

export interface ClientCapabilities {
  rendering: 'kitty-k' | 'kitty-o' | 'iterm2' | 'sixel' | 'truecolor' | '256' | '16' | 'braille';
  cols: number;
  rows: number;
  cellPx: { w: number; h: number };
  bracketedPaste: boolean;
  mouse: boolean;
}

export interface Hello {
  kind: 'hello';
  token: string;
  clientVersion: string;
  protoVersion: number;
  capabilities: ClientCapabilities;
}

export interface HelloAck {
  kind: 'helloAck';
  daemonVersion: string;
  protoVersion: number;
  bootId: string;
  sessionId: string;
  subscriptions: string[];
  /**
   * Initial world model delivered inline so the client can begin rendering
   * before any `world.snapshot` event lands. Stub until Day 5+ wires real
   * layout/asset/agent state.
   */
  world: WorldSnapshot;
}

/**
 * Stub WorldSnapshot. The real `layout` shape is `OfficeLayout` from
 * `webview-ui/src/office/types.ts`; the daemon doesn't depend on that type
 * directly, so the layout slot is left as a loose record.
 */
export interface WorldSnapshot {
  schemaVersion: 1;
  worldSeed: number;
  layout: Record<string, unknown> | null;
  assets: {
    /** Flat furniture catalog — clients build rotation/state groups client-side. */
    catalog: FurnitureCatalogEntry[];
    /** Counts only; actual pixel data is fetched via assets.requestBlob. */
    characterCount: number;
    floorCount: number;
    wallCount: number;
  };
  agents: [];
}

/** Wire-safe subset of FurnitureAsset — matches shared/assets/manifestUtils FurnitureAsset. */
export interface FurnitureCatalogEntry {
  id: string;
  name: string;
  label: string;
  category: string;
  file: string;
  width: number;
  height: number;
  footprintW: number;
  footprintH: number;
  isDesk: boolean;
  canPlaceOnWalls: boolean;
  canPlaceOnSurfaces?: boolean;
  backgroundTiles?: number;
  groupId?: string;
  orientation?: string;
  state?: string;
  mirrorSide?: boolean;
  rotationScheme?: string;
  animationGroup?: string;
  frame?: number;
}

/** Type guard: minimal validation that `msg` is shaped like a Hello. */
export function isHello(msg: unknown): msg is Hello {
  if (typeof msg !== 'object' || msg === null) return false;
  const m = msg as Partial<Hello>;
  return (
    m.kind === 'hello' &&
    typeof m.token === 'string' &&
    typeof m.clientVersion === 'string' &&
    typeof m.protoVersion === 'number'
  );
}

export function isReq(msg: unknown): msg is Req {
  if (typeof msg !== 'object' || msg === null) return false;
  const m = msg as Partial<Req>;
  return m.kind === 'req' && typeof m.reqId === 'number' && typeof m.method === 'string';
}

File: client/src/daemon/wire.rs
#![allow(dead_code)]

// Wire protocol types — mirrors daemon/src/rpc/wire.ts (protoVersion = 1).
// Binary framing: tag byte + payload.
//   0x00 NDJSON : [0x00][json bytes][0x0a]
//   0x01 PTY out: [0x01][streamId:u32be][len:u32be][bytes]
//   0x02 asset  : [0x02][assetId:u32be][tier:u8][len:u32be][bytes]
//   0x03 PTY in : [0x03][streamId:u32be][len:u32be][bytes]

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const PROTO_VERSION: u32 = 1;
pub const CLIENT_VERSION: &str = "0.1.0";

// ── Handshake ─────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize)]
pub struct Hello {
    pub kind: &'static str, // "hello"
    pub token: String,
    #[serde(rename = "clientVersion")]
    pub client_version: &'static str,
    #[serde(rename = "protoVersion")]
    pub proto_version: u32,
    pub capabilities: ClientCapabilities,
}

impl Hello {
    pub fn new(token: String, capabilities: ClientCapabilities) -> Self {
        Self {
            kind: "hello",
            token,
            client_version: CLIENT_VERSION,
            proto_version: PROTO_VERSION,
            capabilities,
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct HelloAck {
    #[serde(rename = "bootId")]
    pub boot_id: String,
    #[serde(rename = "daemonVersion")]
    pub daemon_version: String,
    #[serde(rename = "sessionId")]
    pub session_id: String,
    pub world: Value, // WorldSnapshot — parsed lazily
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientCapabilities {
    pub rendering: RenderingCap,
    pub cols: u16,
    pub rows: u16,
    #[serde(rename = "cellPx")]
    pub cell_px: CellPx,
    #[serde(rename = "bracketedPaste")]
    pub bracketed_paste: bool,
    pub mouse: bool,
    #[serde(rename = "sixelCols", skip_serializing_if = "Option::is_none")]
    pub sixel_cols: Option<u16>,
    #[serde(rename = "sixelRows", skip_serializing_if = "Option::is_none")]
    pub sixel_rows: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CellPx {
    pub w: u16,
    pub h: u16,
}

impl ClientCapabilities {
    pub fn stub(cols: u16, rows: u16) -> Self {
        Self {
            rendering: RenderingCap::Truecolor,
            cols,
            rows,
            cell_px: CellPx { w: 8, h: 16 },
            bracketed_paste: true,
            mouse: true,
            sixel_cols: None,
            sixel_rows: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RenderingCap {
    KittyK,
    KittyO,
    Iterm2,
    Sixel,
    Truecolor,
    #[serde(rename = "256")]
    C256,
    #[serde(rename = "16")]
    C16,
    Braille,
}

// ── RPC ───────────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize)]
pub struct Req {
    pub kind: &'static str, // "req"
    #[serde(rename = "reqId")]
    pub req_id: u32,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

#[derive(Debug, Deserialize)]
pub struct Res {
    #[serde(rename = "reqId")]
    pub req_id: u32,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<WireError>,
}

#[derive(Debug, Deserialize)]
pub struct WireError {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Deserialize)]
pub struct Evt {
    pub topic: String,
    pub seq: u64,
    pub ts: u64,
    pub data: Value,
}

#[derive(Debug, Deserialize)]
pub struct Fatal {
    pub code: String,
    pub message: String,
}

// ── Inbound envelope ─────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum Inbound {
    HelloAck(HelloAck),
    Res(Res),
    Evt(Evt),
    Fatal(Fatal),
}
